### PR TITLE
Sort and remove duplicate when set "taskrun.status.taskResults"

### PR DESCRIPTION
**Pull request by vincent-pli**
Fix issue: #2466
1. Sort `pod.containerStatuses` before parse
2. Overwrite older items

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: pkg/reconciler/taskrun/taskrun.go [hunk 1]
```diff
@@ -603,17 +603,19 @@ func (c *Reconciler) createPod(tr *v1beta1.TaskRun, rtr *resources.ResolvedTaskR
...
+		for idx, cs := range pod.Status.ContainerStatuses {
 			if cs.State.Terminated != nil {
 				msg := cs.State.Terminated.Message
 				r, err := termination.ParseMessage(msg)
 				if err != nil {
 					return fmt.Errorf("parsing message for container status %d: %v", idx, err)
 				}
 				taskResults, pipelineResourceResults := getResults(r)
-				taskRun.Status.TaskRunResults = append(taskRun.Status.TaskRunResults, taskResults...)
+				taskRun.Status.TaskRunResults = appendIfMissingOrReplaced(taskRun.Status.TaskRunResults, taskResults)
```
Comment (sbwsg) [inline pkg/reconciler/taskrun/taskrun.go:618]: It was a little tricky to follow the logic of `appendIfMissingOrReplaced` for a couple reasons:
- There are multiple nested loops there while we're also inside this container statuses loop. Lots to keep in my head.
- The use of the slice syntax `[:]` to remove duplicates in the newTaskResults list
- Returning the "old" list at the end

Is there an advantage to performing the dedupe process here on every iteration? Could we reduce the complexity a bit by deduplicating just once after iterating through the sorted container statuses?  Something like this:

```go
  uniq := make([]v1beta1.TaskRunResult, 0)
  latest := make(map[string]v1beta1.TaskRunResult, 0)
  for _, res := range taskRun.Status.TaskRunResults {
    if _, seen := latest[res.Name]; !seen {
      uniq = append(uniq, res)
    }
    latest[res.Name] = res
  }
  for i, res := range uniq {
    uniq[i] = latest[res.Name]
  }
  taskRun.Status.TaskRunResults = uniq
```

It's a couple more lines but less nested and (I think) a tad less confusing than the multiple `appends` / slicing / old vs new.

No problem if that's not preferable for some reason, just highlighting a difficulty I had. Maybe rename `appendIfMissingOrReplaced` to `replaceOrAppend`?

**Comment by sbwsg** [comment]
Thanks for this @vincent-pli could you remove the `Release Notes` section in the PR description if there's none to include?

File: pkg/reconciler/taskrun/taskrun.go (hunk 1 above)
Comment (vincent-pli) [reply pkg/reconciler/taskrun/taskrun.go]: @sbwsg I like your solution, thanks very much.

**Comment by vincent-pli** [comment]
@sbwsg removed the `Release Notes`, thanks

File: pkg/reconciler/taskrun/taskrun.go [hunk 2]
```diff
@@ -603,17 +603,19 @@ func (c *Reconciler) createPod(tr *v1beta1.TaskRun, rtr *resources.ResolvedTaskR
...
+		for idx, cs := range pod.Status.ContainerStatuses {
 			if cs.State.Terminated != nil {
 				msg := cs.State.Terminated.Message
 				r, err := termination.ParseMessage(msg)
 				if err != nil {
 					return fmt.Errorf("parsing message for container status %d: %v", idx, err)
 				}
 				taskResults, pipelineResourceResults := getResults(r)
-				taskRun.Status.TaskRunResults = append(taskRun.Status.TaskRunResults, taskResults...)
+				taskRun.Status.TaskRunResults = replaceOrAppend(taskRun.Status.TaskRunResults, taskResults)
```
Comment (sbwsg) [inline pkg/reconciler/taskrun/taskrun.go:618]: I think we only need to do this once, after the loop, with the list of all results.  After line 621, something like:

```go
taskRun.Status.TaskRunResults = removeDuplicateResults(taskRun.Status.TaskRunResults)
```

(With `func replaceOrAppend` changed to `func removeDuplicateResults(results []v1beta1.TaskRunResult) []v1beta1.TaskRunResult`)

Does that make sense?  Once we've added all the duplicates to the list we can remove them in a single call.

**Comment by sbwsg** [comment]
Looks good, let me know if you want to make the final suggested change and then I'll approve / lgtm!

File: pkg/reconciler/taskrun/taskrun.go (hunk 2 above)
Comment (vincent-pli) [reply pkg/reconciler/taskrun/taskrun.go]: Thanks @sbwsg , I think it's make sense.

File: pkg/reconciler/taskrun/taskrun_test.go [hunk 3]
```diff
@@ -2014,19 +2020,21 @@ func TestUpdateTaskRunResult2(t *testing.T) {
...
-			ContainerStatuses: []corev1.ContainerStatus{{
-				State: corev1.ContainerState{
-					Terminated: &corev1.ContainerStateTerminated{
-						Message: `[{"key":"resultNameOne","value":"resultValueOne", "type": "TaskRunResult"},{"key":"resultNameTwo","value":"resultValueTwo", "type": "TaskRunResult"}]`,
+		pod: corev1.Pod{
+			Status: corev1.PodStatus{
+				ContainerStatuses: []corev1.ContainerStatus{{
+					State: corev1.ContainerState{
+						Terminated: &corev1.ContainerStateTerminated{
+							Message: `[{"key":"resultNameOne","value":"resultValueOne", "type": "TaskRunResult"},{"key":"resultNameTwo","value":"resultValueTwo", "type": "TaskRunResult"}]`,
```
Comment (pritidesai) [inline pkg/reconciler/taskrun/taskrun_test.go:2033]: Can we add a test with duplicate keys here like you have in the issue to make sure that they are acutally deduplicated before updating the status `taskRun.Status.TaskRunResults`? feel free to ignore it if its not applicable ...

Comment (vincent-pli) [reply pkg/reconciler/taskrun/taskrun_test.go:2033]: @pritidesai Make sense, added ,thanks.

**Comment by pritidesai** [comment]
thanks @vincent-pli 
/lgtm
